holoviews/plotting/renderer.py

```diff
@@ -276,16 +285,37 @@ def html(self, obj, fmt=None, css=None, comm=True, **kwargs):
         (mime_type, tag) = MIME_TYPES[fmt], HTML_TAGS[fmt]
         src = HTML_TAGS['base64'].format(mime_type=mime_type, b64=b64)
         html = tag.format(src=src, mime_type=mime_type, css=css)
-        if comm and plot.comm is not None:
-            comm, msg_handler = self.comms[self.mode]
-            if msg_handler is None:
-                return html
-            msg_handler = msg_handler.format(comm_id=plot.comm.id)
-            return comm.template.format(init_frame=html,
-                                        msg_handler=msg_handler,
-                                        comm_id=plot.comm.id)
+        return html
+
+
+    def components(self, obj, fmt=None, comm=True, **kwargs):
+        if isinstance(obj, (Plot, NdWidget)):
+            plot = obj
+        else:
+            plot, fmt = self._validate(obj, fmt)
+
+        data, metadata = {}, {}
+        if isinstance(plot, NdWidget):
+            js, html = plot()
+            plot_id = plot.plot_id
         else:
-            return html
+            html, js = self._figure_data(plot, as_script=True, **kwargs)
+            plot_id = plot.id
+            if comm and plot.comm is not None and self.comm_msg_handler:
+                msg_handler = self.comm_msg_handler.format(plot_id=plot_id)
+                html = plot.comm.html_template.format(init_frame=html,
+                                                      plot_id=plot_id)
+                js += plot.comm.js_template.format(msg_handler=msg_handler,
```

Comment:
Simply concatenating the JS is a little risky if you don't have newlines at the transition.

Reply:
Good point, worth fixing.